File: day1/1_4__dijkstra_quizes/codes/1/main.py
import sys
import heapq

# sys.stdin = open("input.txt", "r")
N, M = map(int, sys.stdin.readline().split())
dist = [float('inf')] * (N + 1)
v = [[] for _ in range(N + 1)]
for _ in range(M):
    from_node, to_node, cost = map(int, sys.stdin.readline().split())
    v[from_node].append((cost, to_node))


def dijkstra(st):
    hq = []
    heapq.heappush(hq, (0, st))
    dist[st] = 0
    while hq:
        cost, num = heapq.heappop(hq)
        if dist[num] < cost:
            continue
        for next_cost, next_num in v[num]:
            next_cost = cost + next_cost
            if next_cost < dist[next_num]:
                dist[next_num] = next_cost
                heapq.heappush(hq, (next_cost, next_num))

dijkstra(0)

if dist[N - 1] == float('inf'):
    print("impossible")
else:
    print(dist[N - 1])

File: day1/1_4__dijkstra_quizes/codes/1/main.cpp
#define _CRT_SECURE_NO_WARNINGS
#include <iostream>
#include <vector>
#include <queue>
#include <climits>

using namespace std;

struct Edge
{
    int num;
    long long cost;
};

struct cmp
{
    bool operator()(Edge a, Edge b)
    {
        return a.cost > b.cost;
    }
};

int N, M;
long long dist[20001];
vector<Edge> v[20001];

void dijkstra(int st)
{
    priority_queue<Edge, vector<Edge>, cmp> pq;
    pq.push({ st, 0 });
    dist[st] = 0;
    while (!pq.empty())
    {
        Edge now = pq.top();
        pq.pop();

        // 가지치기 조건
        if (dist[now.num] < now.cost)
            continue;

        for (int i = 0; i < v[now.num].size(); i++)
        {
            long long nextCost = now.cost + v[now.num][i].cost;

            int tar = v[now.num][i].num;
            if (nextCost < dist[tar])
            {
                dist[tar] = nextCost;
                pq.push({ tar, nextCost });
            }
        }
    }
}

int main()
{
    // freopen("input.txt", "r", stdin);
    cin >> N >> M;
    int from, to;
    long long cost;
    for (int i = 0; i < M; i++)
    {
        cin >> from >> to >> cost;
        v[from].push_back({ to, cost });
    }

    // 초기 dist 세팅
    for (int i = 1; i <= N; i++)
    {
        dist[i] = LLONG_MAX;
    }

    dijkstra(0);

    if (dist[N - 1] == LLONG_MAX) {
      cout << "impossible\n";
    } else {
      cout << dist[N - 1] << "\n";
    }
}
